rosidl_runtime_c/src/type_description_utils.c

```diff
@@ -0,0 +1,1569 @@
+          "Partial in-place replacements might have already happened.");
+        return RCUTILS_RET_ERROR;
+      }
+
+      ret = rosidl_runtime_c__String__assign(&(field->type.nested_type_name), repl);
+      allocator.deallocate(repl, allocator.state);
+      if (!ret) {
+        RCUTILS_SET_ERROR_MSG(
+          "Could not assign individual type description field nested type name. Beware: "
+          "Partial in-place replacements might have already happened.");
+        return RCUTILS_RET_ERROR;
+      }
+    }
+  }
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rosidl_runtime_c_type_description_utils_repl_all_type_description_type_names_in_place(
+  rosidl_runtime_c__type_description__TypeDescription * type_description,
+  const char * from,
+  const char * to)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(type_description, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(from, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(to, RCUTILS_RET_INVALID_ARGUMENT);
+
+  rcutils_ret_t ret = RCUTILS_RET_ERROR;
+
+  /* *INDENT-OFF* */  // Uncrustify will dislodge the NOLINT
+  ret = rosidl_runtime_c_type_description_utils_repl_individual_type_description_type_names_in_place(  // NOLINT
+    &type_description->type_description, from, to);
+  /* *INDENT-ON* */
+  if (ret != RCUTILS_RET_OK) {
+    rcutils_error_string_t error_string = rcutils_get_error_string();
+    rcutils_reset_error();
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Could not replace main type description type name:\n%s", error_string.str);
+    return ret;
+  }
+
+  if (type_description->referenced_type_descriptions.data) {
+    for (size_t i = 0; i < type_description->referenced_type_descriptions.size; ++i) {
+      rosidl_runtime_c__type_description__IndividualTypeDescription * individual_type_description =
+        &type_description->referenced_type_descriptions.data[i];
+
+      /* *INDENT-OFF* */  // Uncrustify will dislodge the NOLINT
+      ret = rosidl_runtime_c_type_description_utils_repl_individual_type_description_type_names_in_place(  // NOLINT
+        individual_type_description, from, to);
+      /* *INDENT-ON* */
+
+      if (ret != RCUTILS_RET_OK) {
+        rcutils_error_string_t error_string = rcutils_get_error_string();
+        rcutils_reset_error();
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Could not replace type names in referenced type. Beware: "
+          "Partial in-place replacements might have already happened:\n%s",
+          error_string.str);
+        return ret;
+      }
+    }
+  }
+  return RCUTILS_RET_OK;
+}
+
+
+// =================================================================================================
+// DESCRIPTION PRINTING
+// =================================================================================================
+void
+rosidl_runtime_c_type_description_utils_print_field_type(
```

Comment:
We've discussed this and we're going to remove these print functions before merging this because:

- we should not print directly, if anything we should to_string and let the user printf
- the user can and should do this themselves, so they can handle indention and formatting themselves
- the user can already use the ROS message provided functions to convert this to yaml